unit_tests/memory_monitor/test_memory_monitor_main.cpp

```diff
@@ -0,0 +1,297 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+#define CATCH_CONFIG_MAIN
+#include "catch2/catch.hpp"
+
+#include <cstdlib>
+#include <malloc.h>
+#include <thread>
+
+#include "memory_monitor_sr.hpp"
+
+// For multithread test
+static const int num_threads = 100;
+
+// Global variable for whole unit-test
+char *test_mem_in_the_scope = NULL;
+char *test_mem_in_the_scope_with_src = NULL;
+char **test_mem_in_the_scope_multithread = NULL;
+char *test_mem_out_of_scope = NULL;
+char *test_mem_small = NULL;
+
+using namespace cubmem;
+
+// Global memory_monitor
+memory_monitor *test_mmon_Gl = nullptr;
+
+// Dummy APIs
+bool mmon_is_memory_monitor_enabled ()
+{
+  return (test_mmon_Gl != nullptr);
+}
+
+int mmon_initialize (const char *server_name)
+{
+  test_mmon_Gl = new (std::nothrow) memory_monitor ("unittest");
+  if (test_mmon_Gl == nullptr)
+    {
+      fprintf (stdout, "memory_monitor allocation failed\n");
+    }
+  return 0;
+}
+
+void mmon_finalize ()
+{
+  if (mmon_is_memory_monitor_enabled ())
+    {
+      delete test_mmon_Gl;
+      test_mmon_Gl = nullptr;
+    }
+}
+
+size_t mmon_get_allocated_size (char *ptr)
+{
+  return test_mmon_Gl->get_allocated_size (ptr);
+}
+
+void mmon_add_stat (char *ptr, const size_t size, const char *file, const int line)
+{
+  test_mmon_Gl->add_stat (ptr, size, file, line);
+}
+
+void mmon_sub_stat (char *ptr)
+{
+  test_mmon_Gl->sub_stat (ptr);
+}
+
+void mmon_aggregate_server_info (MMON_SERVER_INFO &server_info)
+{
+  test_mmon_Gl->aggregate_server_info (server_info);
+}
+// Dummy APIs end
+
+// Find target stat
+int find_test_stat (MMON_SERVER_INFO &server_info, std::string target)
+{
+  int cnt = 0;
+  for (const auto &[stat_name, mem_usage] : server_info.stat_info)
+    {
+      if (stat_name.compare (target) == 0)
+	{
+	  return cnt;
+	}
+      cnt++;
+    }
+
+  // there is no matching target
+  return -1;
+}
+
+/* ***IMPORTANT!!***
+ * The unit tests for the 'memory_monitor' class are intertwined within one cohesive framework.
+ * Each test depends on all the previous tests. These tests can be broadly categorized into
+ * two categories: 'add' and 'sub'. When developers add new tests,
+ * they must consider the entire test suite.
+ */
+
+// Test add_stat()
+//
+// Test add_stat() with single thread when different file paths
+// that will be made same stat name are given
+//
+// Expected Result: stat name "add_test.c:100" with value "allocated_size + allocated_size_2"
+TEST_CASE ("Test mmon_add_stat", "")
```

Comment:
I think one more case is required.

This scenario seems like an abnormal case. How about adding the normal operation case first? 

and then when adding the abnormal case, additional explanation may be needed.
For the abnormal case explanation, I think it is enough to summarize it as "an issue where the module fails to distinguish different path names".

Reply:
I agree.